### ISSUE TITLE
Custom IMAP command

### ISSUE THREAD

**Issue by libec** [MailCore/mailcore2#1247]
Hi,

I need to send custom IMAP command in my app (let's say ACTIVATE_CUSTOM_SERVICE) so my communication would go like:

1 STARTTLS
2 CAPABILITY
3 ACTIVATE_CUSTOM_SERVICE

Is there any way to do so? Is there an API to create custom operation which would result in sending this custom command to the server? Thanks.

**Comment by dinhviethoa** [comment]
The main issue would be to parse the result of the custom API.
Could you tell me more about this custom command?

**Comment by libec** [comment]
I'm supposed to call ACTIVATE_CUSTOM_SERVICE STATE=START when I want the IMAP server to send me additional data to display to user. Then when the user no longer applies for the service I'm gonna have to send ACTIVATE_CUSTOM_SERVICE STATE=STOP. Server's gonna respond with OK Service is not <STATE> or NO Service is now <additional possible states>. So the response should not be that hard to parse. Can you point me towards solving this? Thanks a lot.

**Comment by libec** [comment]
It might even help if you pointed me how you create the FETCH message. I would then be able to duplicate it to create my own command I think. Any help is appreciated, thanks.

**Comment by dinhviethoa** [comment]
You would have to create a command on libetpan project (https://github.com/dinhviethoa/libetpan), then in src/core/imap/MCIMAPSession.cpp, create the method and then create the operation in MCAsyncIMAPConnection/MCAsyncIMAPSession.
Let me know if shows a clear path for you.

**Comment by libec** [comment]
Thanks a lot, I can work with this. Do you want me to do it so you can specify the command at the top layer so you can use it in the libetpan and MailCore2 (similar as selecting folder from which to fetch messages) or you does it not make sense outside my project?

**Comment by dinhviethoa** [comment]
I think it could be something like:
int mailimap_custom_command(mailimap * imap, const char * command);

**Comment by libec** [comment]
Okay, I've been able to make changes in **libetpan**, now I need to use my fork of libetpan with my fork of **mailcore**. Is there anything else I need to do than change get-ios.sh and build-libetpan-ios.sh files? I changed `get_prebuild_dep` to `build_git_ios` in `get-ios.sh` and revision and url for my own fork in `build-libetpan-ios.sh`. But Xcode keeps crashing on the get-ios script i guess, see the log:

working in /Users/libor/MailCore-Builds/dependencies/workdir/20151019135651
fatal: repository 'ctemplate-ios' does not exist
/Users/libor/Development/iOS/mailcore2/scripts/include.sh/build-dep.sh: line 49: cd: ctemplate-ios: No such file or directory
cp: /Users/libor/MailCore-Builds/dependencies/downloads/ctemplate-ios: No such file or directory
/Users/libor/Development/iOS/mailcore2/scripts/include.sh/build-dep.sh: line 58: cd: /Users/libor/MailCore-Builds/dependencies/workdir/20151019135651/src/ctemplate-ios: No such file or directory
building ctemplate-ios 3 -
/Users/libor/Development/iOS/mailcore2/scripts/include.sh/build-dep.sh: line 67: cd: /Users/libor/MailCore-Builds/dependencies/workdir/20151019135651/src/ctemplate-ios/build-mac: No such file or directory
building iphoneos8.4
/Users/libor/Development/iOS/mailcore2/scripts/include.sh/build-dep.sh: line 70: xctool: command not found
failed

 Thanks for your help on this.

**Comment by dinhviethoa** [comment]
Could you install xctool with brew?

```
$ brew install xctool
```

If you don't have brew, install it through http://brew.sh/

**Comment by libec** [comment]
Seems odd, I have xctool installed already. I'm gonna work on this further

**Comment by dinhviethoa** [comment]
An other option could be to build the dependencies manually and put them in mailcore2/Externals.

**Comment by libec** [comment]
Yeah, that should do for now, I'll give this a go, thanks

**Comment by libec** [comment]
So I finally managed to get libetpan and mailcore working in my project with the changes I made so I can send the command ✨

What do you suggest would be the best way to implement the custom command operation completion handler? String with server response seems kinda cumbersome. I'd like to finish this so others can use it as well.

**Comment by dinhviethoa** [comment]
IMAPSession::login() is checking the value of mImap->imap_response, which is the content of the IMAP response. For example, if you service reply `OK Service is now 'STATE'`, the value of mImap->imap_response will be `Service is now 'STATE'`. Does it help you shape a patch?

**Comment by libec** [comment]
Yes, this seems good enough. Thanks a lot

**Comment by libec** [comment]
I pull requested libetpan first, if you merge it, I'm gonna continue with mailcore :+1:

**Comment by dinhviethoa** [comment]
Sounds like a good plan.

**Comment by dinhviethoa** [comment]
binary is updated. You need to update your patch with the changes of master.
